docs/DeveloperGuide.md

```diff
@@ -1,25 +1,277 @@
+question and answer fields of a flashcard for the user. Without this
+feature, long question and answer fields will follow the default
+wrapping style of the console. When words are truncated unnecessarily,
+it is going to be distracting and annoying for students trying to
+study. We illustrate the problem in the following example:
+```
+This is a long question (or maybe answer) field. Suppose tha
+t our console is 60 characters wide, we see that the word "t
+hat" was truncated in the first line and again in the second
+line.
+```
+In this section, we define the following terms:
+* `lineLength` is the maximum number of characters on a line,
+set to be equal to Ui.DASH_LINES.length(). This is also the number of
+characters between the start of line and end of line.
+* `label` can be "Question: " or "Answer:   " and is used to indicate
+whether a field is the question or answer of the flashcard.
+* `usableLength` is the number of characters that can be used for
+printing a field. This is also the number of characters between the end
+of `label` and end of line.
+
+The following sequence diagram illustrates the call to the
+`toString(boolean isQuestion, int offset)` method of a `FlashCard`
+object.
+
+![DG-Implementation-Features-PP-Sequence](./images-dg/PP-Sequence.png?raw=true)
+
+The `offset` parameter specifies the number of characters already
+printed on the line before the flashcard field will be printed.
+The `offset` is used by the `formatResponse()` method to determine
+`usableLength`.
+
+`formatResponse()` places as many words as possible on each line until
+the next word does not fit within the `usableLength` of the current
+line. This word is therefore placed on the next line and the process
+repeats until all the words have been formatted into the response. If
+the  length of a single word exceeds the `usableLength`, the word is
+split across multiple lines to prevent the program from looping
+infinitely as it would never be able to fit the word on any line.
+
+Take note that infinite loops can still occur if
+* formatResponse() is called with offset >= `lineLength` or
+* toString(boolean, int) is called with offset >= `lineLength` - length
+of `label`
+
+#### Design Consideration:
+
+In order to maximize `usableLength`,
+`toString(boolean isQuestion, int offset)` is designed to take
+parameter `offset` instead of hardcoding `offset` to be
+`"2147483647. ".length()` which is the maximum possible index when
+listing flashcards. As a result, flashcards with different number of
+digits in the index will be misaligned when listing flashcards.
+
+### Tags for grouping and searching decks
+The purpose of this feature is to provide a mean to group the decks based on their subjects
+and search for relevant decks related to one or more relevant subjects in a robust way. Each created deck will
+tagged to their respective field.
+
+![DG-Implementation-Features-TagArchitecture](./images-dg/TagFeature-UML.png?raw=true)
```

Comment:
![image](https://user-images.githubusercontent.com/50763491/97543877-cf4baf80-1a03-11eb-9959-ff7f2b37e86a.png)
minor issue but the overlapping arrows here make it seems a bit confusing. can be nicer